File: system/ulib/mx/eventpair.cc
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <mx/eventpair.h>

#include <magenta/syscalls.h>

namespace mx {

mx_status_t eventpair::create(eventpair* endpoint0, eventpair* endpoint1,
                              uint32_t flags) {
    mx_handle_t h[2];
    mx_status_t result = mx_eventpair_create(h, flags);
    endpoint0->reset(h[0]);
    endpoint1->reset(h[1]);
    return result;
}

} // namespace mx

File: public/lib/fidl/dart/lib/src/core/eventpair.dart
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

part of core;

// TODO: can this class be retired in favour of HandlePairResult?
class Eventpair {
  factory Eventpair() {
    HandlePairResult result = System.eventpairCreate();

    return new Eventpair._(
      status: result.status,
      endpoint0: result.first,
      endpoint1: result.second,
    );
  }

  Eventpair._({
    this.status: ZX.OK,
    this.endpoint0,
    this.endpoint1,
  });

  final int status;
  Handle endpoint0;
  Handle endpoint1;

  Handle passEndpoint0() {
    final Handle result = endpoint0;
    endpoint0 = null;
    return result;
  }

  Handle passEndpoint1() {
    final Handle result = endpoint1;
    endpoint1 = null;
    return result;
  }
}
